How many late days are you using for this assignment? 0 Late Days. 

In [28]:
# Imports and pip installations (if needed) 
%pip install pandas numpy scikit-learn tabulate
import pandas as pd
import numpy as np
from sklearn.preprocessing import OneHotEncoder, StandardScaler




[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: python3 -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


# Part 1: Load the dataset

In [29]:
# Load the given datasets
pd.reset_option('display.max_rows')
pd.reset_option('display.max_columns')
pd.reset_option('display.expand_frame_repr')
categorical_df = pd.read_csv("chronic_kidney_disease_categorical.csv")
numerical_df = pd.read_csv("chronic_kidney_disease_numerical.csv")
#4.2 Assignment
column_mappings = {
    "age": "Age",
    "bp": "Blood Pressure",
    "sg": "Specific Gravity",
    "al": "Albumin",
    "su": "Sugar",
    "rbc": "Red Blood Cells",
    "pc": "Pus Cell",
    "pcc": "Pus Cell Clumps",
    "ba": "Bacteria",
    "bgr": "Blood Glucose Random",
    "bu": "Blood Urea",
    "sc": "Serum Creatinine",
    "sod": "Sodium",
    "pot": "Potassium",
    "hemo": "Hemoglobin",
    "pcv": "Packed Cell Volume",
    "wbcc": "White Blood Cell Count",
    "rbcc": "Red Blood Cell Count",
    "htn": "Hypertension",
    "dm": "Diabetes Mellitus",
    "cad": "Coronary Artery Disease",
    "appet": "Appetite",
    "pe": "Pedal Edema",
    "ane": "Anemia",
}
numerical_df.rename(columns=column_mappings, inplace=True)
categorical_df.rename(columns=column_mappings, inplace=True)
# Print the data
print("Categorical Data:")
print(categorical_df.head())
print(categorical_df.tail())
print("\nNumerical Data:")
print(numerical_df.head())
print(numerical_df.tail())

Categorical Data:
   unique_id  Albumin  Sugar Red Blood Cells  Pus Cell Pus Cell Clumps  \
0     224481      3.0    NaN             NaN       NaN      notpresent   
1     992643      0.0    0.0             NaN       NaN      notpresent   
2     308740      1.0    0.0        abnormal  abnormal      notpresent   
3     450314      1.0    0.0             NaN    normal      notpresent   
4     881763      0.0    0.0          normal    normal      notpresent   

     Bacteria Hypertension Diabetes Mellitus Coronary Artery Disease Appetite  \
0  notpresent          yes               yes                      no     good   
1  notpresent           no                no                      no     good   
2  notpresent          yes                no                      no     good   
3  notpresent           no                no                      no     good   
4  notpresent           no                no                      no     good   

  Pedal Edema Anemia  Target  
0         yes     n

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The Chronic Kidney Disease dataset is essentially various clinical records where a patients data is captured and used to predict whether they have Chronic Kidney Disease (this is the label). The features are health indicators, which provide precise, quantifiable data such as lab test results, while others are categorical, representing qualitative assessments. For example, numerical features may include values like the patient’s age or specific test results, and categorical features might indicate the presence or absence of certain conditions or abnormal test findings.
• Numerical examples: Age, Blood Pressure, Serum Creatinine
• Categorical examples: Anemia (ane), Hypertension (htn), Diabetes Mellitus (dm)    

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [30]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_numerical = numerical_df[numerical_df.duplicated()]
print(f"Total duplicate rows in numerical dataset: {len(duplicate_numerical)}")
print(duplicate_numerical)
# Drop these duplicate rows
numerical_df.drop_duplicates(inplace=True)
# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_categorical = categorical_df[categorical_df.duplicated()]
print(f"Total duplicate rows in categorical dataset: {len(duplicate_categorical)}")
print(duplicate_categorical)
# Drop these duplicate rows
categorical_df.drop_duplicates(inplace=True)



Total duplicate rows in numerical dataset: 30
     unique_id   Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
107     872034  70.0            90.0                 184.0        98.6   
149     721270  58.0            70.0                 102.0        48.0   
171     240975  47.0            80.0                  95.0        35.0   
173     441170  79.0            80.0                 111.0        44.0   
185     261137  35.0            60.0                 105.0        39.0   
194     769512  38.0            60.0                  91.0        36.0   
213     354223  64.0            70.0                  97.0        27.0   
244     302906  59.0           100.0                   NaN        96.0   
245     431871  57.0            60.0                 132.0        18.0   
248     687021  52.0            80.0                  99.0        25.0   
249     843449  30.0            80.0                 131.0        38.0   
265     628271  65.0            90.0                 270.0        

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, not all Unique Id's are present in both datasets. The rows that are missing from the numerical dataset would simply present in the merged table as "NaN" for all values that were supposed to be numerical. For example: 47 208034 NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN notpresent notpresent yes yes yes good yes no ckd, where ALL NaN values are from the numerical dataset, where unique id 208034 is not present. 

In [31]:

# Merge the two given numerical and categorical datasets based on their unique_ID.
if 'unique_id' in categorical_df.columns and 'unique_id' in numerical_df.columns:
    merged_df = pd.merge(numerical_df, categorical_df, on="unique_id", how="outer")
else:
    raise ValueError("unique_id column is missing from one of the datasets")
#Print the combined dataset
print(merged_df.head()) 
print(merged_df.tail()) 



   unique_id   Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0     102677  59.0            70.0                  76.0       186.0   
1     103570  24.0            80.0                 125.0         NaN   
2     104675  65.0            80.0                 215.0       133.0   
3     105923  90.0            90.0                 139.0        89.0   
4     107297  60.0            70.0                   NaN         NaN   

   Serum Creatinine  Sodium  Potassium  Hemoglobin  Packed Cell Volume  ...  \
0              15.0   135.0        7.6         7.1                22.0  ...   
1               NaN   136.0        3.5        15.4                43.0  ...   
2               2.5     NaN        NaN        13.2                41.0  ...   
3               3.0   140.0        4.1        12.0                37.0  ...   
4               NaN     NaN        NaN        16.4                43.0  ...   

   Pus Cell  Pus Cell Clumps    Bacteria  Hypertension Diabetes Mellitus  \
0  abnormal     

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [32]:

# Count rows with at least one missing value
rows_with_missing = merged_df.shape[0] - merged_df.dropna().shape[0]
missing_percentage = (rows_with_missing / merged_df.shape[0]) * 100
# Print before dropping missing values
print(f"Percentage of rows with missing values: {missing_percentage:.2f}%")
print("Before dropping missing values:")
print(merged_df.head()) 
print(merged_df.tail()) 
# Drop rows with missing values
merged_df.dropna(inplace=True)
# Print the dataset after removing missing values
print("After dropping missing values:")
merged_df.reset_index(drop=True, inplace=True)
print(merged_df.head()) 
print(merged_df.tail()) 


Percentage of rows with missing values: 61.75%
Before dropping missing values:
   unique_id   Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0     102677  59.0            70.0                  76.0       186.0   
1     103570  24.0            80.0                 125.0         NaN   
2     104675  65.0            80.0                 215.0       133.0   
3     105923  90.0            90.0                 139.0        89.0   
4     107297  60.0            70.0                   NaN         NaN   

   Serum Creatinine  Sodium  Potassium  Hemoglobin  Packed Cell Volume  ...  \
0              15.0   135.0        7.6         7.1                22.0  ...   
1               NaN   136.0        3.5        15.4                43.0  ...   
2               2.5     NaN        NaN        13.2                41.0  ...   
3               3.0   140.0        4.1        12.0                37.0  ...   
4               NaN     NaN        NaN        16.4                43.0  ...   

   Pus Cell  

## Part 3.4: Sort the dataset according to the Labels

In [33]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
if 'Target' in merged_df.columns:
    merged_df.sort_values(by='Target', inplace=True)
    merged_df.reset_index(drop=True, inplace=True)
else:
    raise ValueError("Target column is missing from the dataset")
# Print the dataset
print(merged_df.head()) 
print(merged_df.tail()) 

   unique_id   Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0     102677  59.0            70.0                  76.0       186.0   
1     337834  48.0            70.0                 117.0        56.0   
2     343710  61.0            80.0                 173.0       148.0   
3     349892  59.0            70.0                 424.0        55.0   
4     353025  40.0            70.0                 253.0       150.0   

   Serum Creatinine  Sodium  Potassium  Hemoglobin  Packed Cell Volume  ...  \
0              15.0   135.0        7.6         7.1                22.0  ...   
1               3.8   111.0        2.5        11.2                32.0  ...   
2               3.9   135.0        5.2         7.7                24.0  ...   
3               1.7   138.0        4.5        12.6                37.0  ...   
4              11.9   132.0        5.6        10.9                31.0  ...   

   Pus Cell  Pus Cell Clumps    Bacteria  Hypertension Diabetes Mellitus  \
0  abnormal     

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [34]:
# Write code here
categorical_columns = ['Red Blood Cells', 'Pus Cell', 'Pus Cell Clumps', 'Bacteria', 'Hypertension', 'Diabetes Mellitus', 'Coronary Artery Disease', 'Appetite', 'Pedal Edema', 'Anemia', 'Target']
merged_df = pd.get_dummies(merged_df, columns=categorical_columns, drop_first=False).astype(int)
# Print the dataset
print(merged_df.head()) 
print(merged_df.tail()) 

   unique_id  Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0     102677   59              70                    76         186   
1     337834   48              70                   117          56   
2     343710   61              80                   173         148   
3     349892   59              70                   424          55   
4     353025   40              70                   253         150   

   Serum Creatinine  Sodium  Potassium  Hemoglobin  Packed Cell Volume  ...  \
0                15     135          7           7                  22  ...   
1                 3     111          2          11                  32  ...   
2                 3     135          5           7                  24  ...   
3                 1     138          4          12                  37  ...   
4                11     132          5          10                  31  ...   

   Coronary Artery Disease_no  Coronary Artery Disease_yes  Appetite_good  \
0                    

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Yes, it would be equivalent and actually take less memory since you can imply one column based on the other, since they can't be both. For example, if we had rbc_normal (true), then you would know that for that row, rbc_abnormal (false) must follow. This wouldn't be possible for something with more than 2 values because you can't imply what value the other one is given a false value in the column. An example with the condiments is if we had one row called c_ketchup, (meaning condiment ketchup) as false, then you wouldn't know whether it was mustard, mayo, ranch, or otherwise. 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [35]:
# Remove outliers
numerical_columns = ['Age', 'Blood Pressure', 'Blood Glucose Random', 'Blood Urea', 'Serum Creatinine', 'Sodium', 'Potassium', 'Hemoglobin', 'Packed Cell Volume', 'White Blood Cell Count', 'Red Blood Cell Count']
for col in numerical_columns:
    mean = merged_df[col].mean()
    std = merged_df[col].std()
    merged_df = merged_df[(merged_df[col] > mean - 3 * std) & (merged_df[col] < mean + 3 * std)]
# Print the dataset
merged_df.reset_index(drop=True, inplace=True)
print(merged_df.head()) 
print(merged_df.tail()) 

   unique_id  Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0     343710   61              80                   173         148   
1     397388   63             100                   280          35   
2     438182   73              80                   253         142   
3     474407   21              90                   107          40   
4     475200   64             100                   163          54   

   Serum Creatinine  Sodium  Potassium  Hemoglobin  Packed Cell Volume  ...  \
0                 3     135          5           7                  24  ...   
1                 3     143          3          13                  40  ...   
2                 4     138          5          10                  33  ...   
3                 1     125          3           8                  23  ...   
4                 7     140          4           7                  26  ...   

   Coronary Artery Disease_no  Coronary Artery Disease_yes  Appetite_good  \
0                    

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [36]:
# Normalize the all Numerical Attributes in the dataset.
scaler = StandardScaler()
merged_df[numerical_columns] = scaler.fit_transform(merged_df[numerical_columns])
# Print the dataset
print(merged_df.head()) 
print(merged_df.tail()) 

   unique_id       Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0     343710  0.810190        0.730553              1.146438    4.014305   
1     397388  0.936367        2.776102              3.520558   -0.225995   
2     438182  1.567253        0.730553              2.921481    3.789156   
3     474407 -1.713353        1.753327             -0.317973   -0.038371   
4     475200  0.999456        2.776102              0.924557    0.486976   

   Serum Creatinine    Sodium  Potassium  Hemoglobin  Packed Cell Volume  ...  \
0          1.359270 -0.959702   1.538050   -2.973247           -2.905503  ...   
1          1.359270  0.448743  -1.101442   -0.408219           -0.590673  ...   
2          2.019062 -0.431535   1.538050   -1.690733           -1.603411  ...   
3          0.039687 -2.720258  -1.101442   -2.545742           -3.050180  ...   
4          3.998437 -0.079424   0.218304   -2.973247           -2.616150  ...   

   Coronary Artery Disease_no  Coronary Artery Disease_y

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: At this point in the calculations, since each row is considered a unique value and we aren't using these values to track patients, there no real need to have unique id's, and instead we can just opt to use the index values provided. Having the ID's are just needless clutter. 

In [37]:
#Remove that column
unnecessary_columns = ['unique_id'] if 'unique_id' in merged_df.columns else []
merged_df.drop(columns=unnecessary_columns, inplace=True)
# Print the dataset
print(merged_df.head()) 
print(merged_df.tail()) 

        Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0  0.810190        0.730553              1.146438    4.014305   
1  0.936367        2.776102              3.520558   -0.225995   
2  1.567253        0.730553              2.921481    3.789156   
3 -1.713353        1.753327             -0.317973   -0.038371   
4  0.999456        2.776102              0.924557    0.486976   

   Serum Creatinine    Sodium  Potassium  Hemoglobin  Packed Cell Volume  \
0          1.359270 -0.959702   1.538050   -2.973247           -2.905503   
1          1.359270  0.448743  -1.101442   -0.408219           -0.590673   
2          2.019062 -0.431535   1.538050   -1.690733           -1.603411   
3          0.039687 -2.720258  -1.101442   -2.545742           -3.050180   
4          3.998437 -0.079424   0.218304   -2.973247           -2.616150   

   White Blood Cell Count  ...  Coronary Artery Disease_no  \
0                0.519658  ...                           0   
1                0.785483  .

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [38]:
# Export the dataframe to a new csv file
merged_df.to_csv("cleaned_kidney_disease_data.csv", index=False)
# Export the dataframe to a new json file
merged_df.to_json("cleaned_kidney_disease_data.json", orient='records')


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

|    |       age |        bp |        bgr |         bu |        sc |        sod |       pot |      hemo |       pcv |       wbcc |      rbcc |   al |   su |   rbc_abnormal |   rbc_normal |   pc_abnormal |   pc_normal |   pcc_notpresent |   pcc_present |   ba_notpresent |   ba_present |   htn_no |   htn_yes |   dm_no |   dm_yes |   cad_no |   cad_yes |   appet_good |   appet_poor |   pe_no |   pe_yes |   ane_no |   ane_yes |   Target_ckd |   Target_notckd |
|---:|----------:|----------:|-----------:|-----------:|----------:|-----------:|----------:|----------:|----------:|-----------:|----------:|-----:|-----:|---------------:|-------------:|--------------:|------------:|-----------------:|--------------:|----------------:|-------------:|---------:|----------:|--------:|---------:|---------:|----------:|-------------:|-------------:|--------:|---------:|---------:|----------:|-------------:|----------------:|
|  0 |  0.81019  |  0.730553 |  1.14644   |  4.0143    | 1.35927   | -0.959702  |  1.53805  | -2.97325  | -2.9055   |  0.519658  | -1.87278  |    2 |    0 |              1 |            0 |             1 |           0 |                1 |             0 |               1 |            0 |        0 |         1 |       0 |        1 |        0 |         1 |            0 |            1 |       0 |        1 |        0 |         1 |            1 |               0 |
|  1 |  0.936367 |  2.7761   |  3.52056   | -0.225995  | 1.35927   |  0.448743  | -1.10144  | -0.408219 | -0.590673 |  0.785483  | -0.724947 |    2 |    2 |              0 |            1 |             0 |           1 |                1 |             0 |               0 |            1 |        0 |         1 |       1 |        0 |        0 |         1 |            1 |            0 |       1 |        0 |        1 |         0 |            1 |               0 |
|  2 |  1.56725  |  0.730553 |  2.92148   |  3.78916   | 2.01906   | -0.431535  |  1.53805  | -1.69073  | -1.60341  | -0.366426  | -0.724947 |    2 |    0 |              1 |            0 |             1 |           0 |                1 |             0 |               1 |            0 |        0 |         1 |       0 |        1 |        0 |         1 |            1 |            0 |       1 |        0 |        1 |         0 |            1 |               0 |
|  3 | -1.71335  |  1.75333  | -0.317973  | -0.0383712 | 0.0396867 | -2.72026   | -1.10144  | -2.54574  | -3.05018  |  1.93739   | -1.87278  |    4 |    0 |              0 |            1 |             1 |           0 |                0 |             1 |               0 |            1 |        1 |         0 |       1 |        0 |        1 |         0 |            1 |            0 |       1 |        0 |        0 |         1 |            1 |               0 |
|  4 |  0.999456 |  2.7761   |  0.924557  |  0.486976  | 3.99844   | -0.0794236 |  0.218304 | -2.97325  | -2.61615  | -0.233513  | -1.87278  |    4 |    2 |              1 |            0 |             1 |           0 |                1 |             0 |               0 |            1 |        0 |         1 |       0 |        1 |        1 |         0 |            1 |            0 |       0 |        1 |        1 |         0 |            1 |               0 |
|  5 |  0.747101 | -1.315    |  3.69806   | -0.18847   | 0.0396867 | -1.83998   | -1.10144  | -2.97325  | -2.76083  |  3.17791   | -1.87278  |    3 |    1 |              0 |            1 |             1 |           0 |                0 |             1 |               1 |            0 |        0 |         1 |       1 |        0 |        1 |         0 |            0 |            1 |       1 |        0 |        0 |         1 |            1 |               0 |
|  6 |  0.494747 |  1.75333  |  0.170164  |  2.47579   | 3.33865   | -1.66392   |  0.218304 | -2.11824  | -2.18212  | -0.720859  | -1.87278  |    2 |    0 |              1 |            0 |             1 |           0 |                1 |             0 |               1 |            0 |        0 |         1 |       1 |        0 |        1 |         0 |            1 |            0 |       1 |        0 |        1 |         0 |            1 |               0 |
|  7 |  1.3149   | -0.292221 |  2.05615   |  2.06302   | 3.33865   | -3.60054   | -1.10144  | -2.11824  | -2.3268   |  1.53865   | -1.87278  |    4 |    3 |              0 |            1 |             1 |           0 |                0 |             1 |               0 |            1 |        0 |         1 |       0 |        1 |        0 |         1 |            1 |            0 |       0 |        1 |        0 |         1 |            1 |               0 |
|  8 | -0.136138 | -1.315    |  0.924557  |  1.91292   | 1.35927   |  0.0966321 |  0.218304 | -2.11824  | -2.3268   |  2.91208   | -1.87278  |    1 |    0 |              0 |            1 |             0 |           1 |                1 |             0 |               1 |            0 |        0 |         1 |       0 |        1 |        1 |         0 |            1 |            0 |       1 |        0 |        1 |         0 |            1 |               0 |
|  9 |  0.873279 | -0.292221 |  0.0148476 |  0.0366784 | 0.0396867 | -0.783646  |  0.218304 | -0.835723 | -0.73535  | -0.0562963 | -1.87278  |    3 |    0 |              0 |            1 |             1 |           0 |                1 |             0 |               1 |            0 |        0 |         1 |       0 |        1 |        1 |         0 |            1 |            0 |       1 |        0 |        1 |         0 |            1 |               0 |
| 10 |  0.431659 |  0.730553 |  2.05615   |  1.19995   | 1.35927   | -0.607591  |  0.218304 | -1.69073  | -1.45873  | -0.277817  | -1.87278  |    3 |    1 |              0 |            1 |             1 |           0 |                0 |             1 |               0 |            1 |        0 |         1 |       0 |        1 |        1 |         0 |            1 |            0 |       0 |        1 |        1 |         0 |            1 |               0 |
| 11 |  0.999456 | -1.315    |  2.61085   |  0.637075  | 2.01906   | -0.607591  |  1.53805  | -2.11824  | -2.18212  | -0.233513  | -1.87278  |    4 |    1 |              1 |            0 |             1 |           0 |                1 |             0 |               0 |            1 |        0 |         1 |       0 |        1 |        1 |         0 |            0 |            1 |       0 |        1 |        1 |         0 |            1 |               0 |
| 12 |  1.44108  | -1.315    | -0.0739045 |  3.15123   | 2.67885   | -0.783646  |  0.218304 | -1.26323  | -1.31406  |  3.17791   | -0.724947 |    4 |    0 |              0 |            1 |             0 |           1 |                1 |             0 |               1 |            0 |        0 |         1 |       0 |        1 |        1 |         0 |            0 |            1 |       0 |        1 |        1 |         0 |            1 |               0 |
| 13 |  1.44108  | -0.292221 |  2.16709   |  1.53767   | 1.35927   | -1.31181   |  0.218304 | -1.69073  | -1.60341  | -1.07529   | -1.87278  |    3 |    0 |              0 |            1 |             1 |           0 |                0 |             1 |               0 |            1 |        0 |         1 |       0 |        1 |        0 |         1 |            1 |            0 |       1 |        0 |        1 |         0 |            1 |               0 |
| 14 |  0.684013 |  0.730553 |  4.03088   | -0.225995  | 0.0396867 | -3.24843   | -1.10144  | -1.69073  | -1.31406  |  1.27283   | -0.724947 |    1 |    0 |              1 |            0 |             0 |           1 |                1 |             0 |               1 |            0 |        1 |         0 |       0 |        1 |        1 |         0 |            0 |            1 |       1 |        0 |        1 |         0 |            1 |               0 |


## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Generally speaking, the most important part is proof-reading the data once its readable. GPT and LLMS in general have alternative outputs that they can choose from which might be able to fit your ideal outcome. Alternatively, you can use an LLM that is explicitly designed for this type of output, such as an AI that is meant to convert plaintext to markdown tables. 

In [ ]:
# Code to rename all the columns in the dataset
column_mappings = {
    "age": "Age",
    "bp": "Blood Pressure",
    "sg": "Specific Gravity",
    "al": "Albumin",
    "su": "Sugar",
    "rbc": "Red Blood Cells",
    "pc": "Pus Cell",
    "pcc": "Pus Cell Clumps",
    "ba": "Bacteria",
    "bgr": "Blood Glucose Random",
    "bu": "Blood Urea",
    "sc": "Serum Creatinine",
    "sod": "Sodium",
    "pot": "Potassium",
    "hemo": "Hemoglobin",
    "pcv": "Packed Cell Volume",
    "wbcc": "White Blood Cell Count",
    "rbcc": "Red Blood Cell Count",
    "htn": "Hypertension",
    "dm": "Diabetes Mellitus",
    "cad": "Coronary Artery Disease",
    "appet": "Appetite",
    "pe": "Pedal Edema",
    "ane": "Anemia",
}
# categorical_df.rename(columns=column_mappings, inplace=True)
# numerical_df.rename(columns=column_mappings, inplace=True)
# As a note to the graders, 383GPT scanned the "additional information 
# section" and 2 were unaccounted for since in that section it was 
# mentioned that the column name was "rc" but in the dataset it was 
# "rbcc" so I changed it in the mapping to "rbcc" to match the dataset. 
# The same was done for "wc" which was "wbcc" in the dataset. 
# Additionally, class was listed as a category but wasn't a column. 
# Class was removed from the column_mappings dictionary and target was 
# added. This code was actually ran in part 1 of the assignment so it 
# would apply to everything. 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [ ]:
#Converted SQL to Pandas code
result = merged_df.groupby('Target_ckd').size().reset_index(name='count')
print(result)
#We were told to change the original code to the target column is 
# now "Target_ckd" so I changed it in the code above.


   Target_ckd  count
0           0    111
1           1     22


*Describe what the above code does here* The code is essentially a nice summary of how many people within the chart had or didn't have Chronic Kidney Disease. It lists out how many did and didn't have it in a 2x3 formatted table. 
